src/include/common/stat_registry.h

```diff
@@ -0,0 +1,227 @@
+  }
+
+  /**
+   * Returns the true JSON representation of the current registry,
+   * i.e. follows all the pointers to the performance counters and reads them as JSON.
+   */
+  json GetTrueJson(json *root) {
+    json output = static_cast<json>("{}"_json);
+
+    for (const auto &item : root->items()) {
+      const auto &key = item.key();
+
+      if (key == RESERVED_PC_PTR_STRING) {
+        auto val = root->at(key).get<uintptr_t>();
+        output = reinterpret_cast<PerformanceCounter *>(val)->ToJson();
+      } else if (key == RESERVED_PC_REGISTRANT_PTR) {
+        /* don't include the registrant in the output */
+      } else {
+        auto val = root->at(key).get<json>();
+        output[key] = GetTrueJson(&val);
+      }
+    }
+
+    return output;
+  }
+
+  /**
+   * Frees all the performance counters stored in the registry.
+   */
+  void FreePerformanceCounters(json *root) {
+    for (const auto &item : root->items()) {
+      const auto &key = item.key();
+
+      if (key == RESERVED_PC_PTR_STRING) {
+        auto val = root->at(key).get<uintptr_t>();
+        delete reinterpret_cast<PerformanceCounter *>(val);
+      } else if (key == RESERVED_PC_REGISTRANT_PTR) {
+        /* don't touch the parent */
+      } else {
+        auto val = root->at(key).get<json>();
+        FreePerformanceCounters(&val);
+      }
+    }
+  }
+
+ public:
+  /**
+   * Registers the given performance counter pc at the module specified.
+   * If the specified module contains a binding which conflicts with pc's name,
+   * pc's name will be modified until the binding succeeds.
+   * @param module_path path to destination module
+   * @param pc performance counter to be registered
+   * @param registrant pointer to instance that is registering this counter
+   */
+  void Register(const std::vector<std::string> &module_path, PerformanceCounter *pc, void *registrant) {
+    json *mod = FindModule(module_path, true);
+
+    // generate a new insertion name
+    std::string insert_name = pc->GetName();
+    uint32_t index = 1;
+    while (mod->find(insert_name) != mod->end()) {
+      insert_name = pc->GetName();
+      insert_name.append(std::to_string(index++));
+    }
+    pc->SetName(insert_name);
+
+    // create if doesn't exist
+    mod->emplace(insert_name, "{}"_json);
+    mod->at(insert_name)[RESERVED_PC_PTR_STRING] = reinterpret_cast<uintptr_t>(pc);
+    mod->at(insert_name)[RESERVED_PC_REGISTRANT_PTR] = reinterpret_cast<uintptr_t>(registrant);
+  }
+
+  /**
+   * Deregister and remove the named performance counter from the module specified.
+   *
+   * @param module_path path to destination module
+   * @param name name of the performance counter to be deregistered
+   * @param free_pc true if the performance counter should be freed, false otherwise
+   * @return true if the performance counter was successfully removed, false otherwise
+   */
+  bool Deregister(const std::vector<std::string> &module_path, const std::string &name, bool free_pc) {
```

Comment:
Do we always want these to go out of scope, or maybe retain them if needed?

Reply:
Retain the removed performance counters? They remain in whatever object they were in, they will just not be tracked by the registry any more.